### PR TITLE
Add asserts to catch unintentional use of generic overloads

### PR DESCRIPTION
**Pull request by jkotas**
Delete unnecessary ifdefs

### CODE REVIEW

**Comment by jkotas** [comment]
This is follow up on #1113 . cc @gfoidl

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 1]
```diff
@@ -3,16 +3,18 @@
...
-#endif
 
 namespace System
 {
     internal static partial class SpanHelpers // .T
     {
         public static int IndexOf<T>(ref T searchSpace, int searchSpaceLength, ref T value, int valueLength) where T : IEquatable<T>
         {
+            // The optimized implementation should be used for these types
+            Debug.Assert(typeof(T) != typeof(byte) && typeof(T) != typeof(char));
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs]: The optimized implementation is based on the size of T, so the assert should be
```suggestion
            Debug.Assert(Unsafe.SizeOf<T>() != sizeof(byte) && Unsafe.SizeOf<T>() != sizeof(char));
```
as `sbyte`, `short`, `ushort` will also take the optimized version.

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 2]
```diff
@@ -3,16 +3,19 @@
...
-#endif
 
 namespace System
 {
     internal static partial class SpanHelpers // .T
     {
         public static int IndexOf<T>(ref T searchSpace, int searchSpaceLength, ref T value, int valueLength) where T : IEquatable<T>
         {
+            // The optimized implementation should be used for these types
+            Debug.Assert(!RuntimeHelpers.IsBitwiseEquatable<T>() || !(Unsafe.SizeOf<T>() == sizeof(byte) || Unsafe.SizeOf<T>() == sizeof(char)));
```
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs]: Interested about this: what can cause non-bitwise equatable when the size is less than pointer?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs]: For example:

```
struct MyOneByteStruct : IEquatable<T>
{ 
    byte _flags;

    bool Equals(T other)
    {
        // The structs are considered equal when second bit of flags is equal
        return (_flags & 2) == (other._flags & 2);
    }
}
```

**Review by stephentoub — APPROVED** [review]

**Pull request merged**